### PR TITLE
Django auto-instrumentation example

### PR DESCRIPTION
**Pull request by sm7x**

### CODE REVIEW

File: signalfx-tracing/signalfx-python-tracing/README.md [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# SignalFX Tracing Library for Python 
+
+This repository contains an example application that utilizes auto-instrumentation for 
+supported python web frameworks and clients. <br/>
+
+For more detailed information about the functionality of the provided tracer and instrumentation configuration, 
+please see the [`signalfx-tracing`](https://github.com/signalfx/signalfx-python-tracing/tree/master/scripts) scripts.
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/README.md]: I'd refer to the top level readme instead of the scripts directly.

File: signalfx-tracing/signalfx-python-tracing/django/README.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+
+### Extra Notes
+1. In this example, we used the SignalFx trace decorator `@trace` to automatically create spans 
+for tracing custom logic. Please see [the tracer decoration documentation](https://github.com/signalfx/signalfx-python-tracing#trace-decorator) 
+for more information. <br/>
+
+2. The `signalfx-tracing` module and this application configuration assume that your Smart Agent
+or Gateway is accepting traces at http://localhost:9080/v1/trace.  If this is not the case,
+you can set the `SIGNALFX_INGEST_URL` environment variable to the desired url to suit your
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/README.md]: `SIGNALFX_ENDPOINT_URL` is now preferred

Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/README.md]: OK, thanks. I'll update it

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/secret_settings.py [hunk 3]
```diff
@@ -0,0 +1 @@
+SECRET_KEY = '&g^)t(v(wwu!$+=jov@tnfa2*@eq9q0^2mwa&$y1$gr_25ne9c'
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/blog/blog/secret_settings.py]: You can define this in settings.py directly.  Having a separate file seems a bit overkill

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+import logging
+import os
+from .secret_settings import SECRET_KEY
+
+# Create and configure logger
+if not os.path.exists('log'):
+    os.makedirs('log')
+logging.basicConfig(level='DEBUG',
+                    filename="log/server_log.log",
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py]: Any reason to log in a file and not stdout for this demo?

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/secret_settings.py (hunk 3 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/blog/secret_settings.py]: That's strange. I actually didn't want it committed to the rep. It was part of the .gitignore file. It probably got committed when the .gitignore file was moved

File: signalfx-tracing/signalfx-python-tracing/django/blog/posts/tests.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from django.test import TestCase
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/blog/posts/tests.py]: Is this file necessary?

File: signalfx-tracing/signalfx-python-tracing/django/client.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+from signalfx_tracing import auto_instrument, create_tracer, trace
+# Important: Needed to propagate traces
+from signalfx_tracing.libraries import requests_config
+
+
+# Create and configure logger
+if not os.path.exists('log'):
+    os.makedirs('log')
+logging.basicConfig(level='DEBUG',
+                    filename="log/client_log.log",
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/client.py]: stdout ok?

File: signalfx-tracing/signalfx-python-tracing/django/client.py [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+# create an instance of the SignalFx tracer
+# and configure auto-instrumentation of the tracing instance
+tracer = create_tracer(service_name='BlogClient')
+auto_instrument(tracer)
+
+
+# Setting requests_config.propagate to 'True'  ensures
+# that the relational information between traces is maintained
+# and forwarded.
+requests_config.propagate = True
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/client.py]: This should come before `auto_instrument`

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py (hunk 4 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py]: To make the stdout less busy. I assume a file will be more useful in actual production (which I am taking as the use case)

File: signalfx-tracing/signalfx-python-tracing/django/client.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    def update_blog(self, blog_id, new_title):
+        url_to_update = self.url + '{}/'.format(blog_id)
+        response = requests.patch(url_to_update,
+                                  json=dict(title=new_title),
+                                  auth=self.session.auth, )
+        logging.info('INFO - Updated status: {}'.format(response.status_code))
+        return response.json()
+
+    @trace
+    def delete_blog(self, blog_id):
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/client.py]: mind checking the status code and returning a boolean?

File: signalfx-tracing/signalfx-python-tracing/django/blog/posts/tests.py (hunk 5 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/posts/tests.py]: That was part of my .gitignore file as well. Need to remove a few files

File: signalfx-tracing/signalfx-python-tracing/django/requirements.txt [hunk 9]
```diff
@@ -0,0 +1,6 @@
+django
+djangorestframework
+coreapi
+pyyaml
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/requirements.txt]: I'm not seeing coreapi or pyyaml used anywhere

File: signalfx-tracing/signalfx-python-tracing/django/client.py (hunk 7 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: OK. It seemed to work when I tested it, but I didn't delve into the traces sent. Will change it.

File: signalfx-tracing/signalfx-python-tracing/django/client.py (hunk 8 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: Sure, can do that

File: signalfx-tracing/signalfx-python-tracing/README.md (hunk 1 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/README.md]: OK

File: signalfx-tracing/signalfx-python-tracing/django/README.md (hunk 2 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/README.md]: Updated

File: signalfx-tracing/signalfx-python-tracing/django/requirements.txt (hunk 9 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/requirements.txt]: localhost:8001/schema

File: signalfx-tracing/signalfx-python-tracing/django/client.py (hunk 7 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: Changed it

File: signalfx-tracing/signalfx-python-tracing/django/blog/posts/tests.py (hunk 5 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/posts/tests.py]: Added to .gitignore and removed

File: signalfx-tracing/signalfx-python-tracing/README.md (hunk 1 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/README.md]: Checked the link again, it links to the README of the `sfx-py-trace-bootstrap` and `sfx-py-trace` scripts (SignalFx-Tracing Library for Python Utilities), isn't that helpful?

File: signalfx-tracing/signalfx-python-tracing/django/client.py (hunk 8 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: Done

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/secret_settings.py (hunk 3 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/blog/secret_settings.py]: Fixed

File: signalfx-tracing/signalfx-python-tracing/django/README.md (hunk 2 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/README.md]: BTW, `SIGNALFX_ENDPOINT_URL` didn't work when I just tried using it for a test run. Do I have to update something in my environment?

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py (hunk 4 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py]: Switched to stdout

File: signalfx-tracing/signalfx-python-tracing/django/client.py (hunk 6 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: switched to stdout

File: signalfx-tracing/signalfx-python-tracing/README.md [hunk 10]
```diff
@@ -0,0 +1,11 @@
+# SignalFX Tracing Library for Python
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/README.md]: `Fx`

File: signalfx-tracing/signalfx-python-tracing/django/client.py [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+        url_to_delete = self.url + '{}/'.format(blog_id)
+        response = requests.delete(url_to_delete,
+                                   auth=self.session.auth, )
+        success = (response.status_code == 204)
+        logging.info('INFO - Deleted status: {}'.format(success))
+
+
+def main():
+    client = BlogClient('http://localhost:8001/articles/', 'admin', 't00c0mm0n')
```
Comment (mpetazzoni) [inline signalfx-tracing/signalfx-python-tracing/django/client.py]: Should this remain here?

Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: Thanks for spotting that (it slipped when I was re-testing). Updated it to 'USERNAME' and 'PASSWORD'

File: signalfx-tracing/signalfx-python-tracing/README.md (hunk 10 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/README.md]: Thanks a bunch! Fixed.

File: signalfx-tracing/signalfx-python-tracing/django/client.py (hunk 11 above)
Comment (mpetazzoni) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: That's fine. Another approach is to make this program take command line arguments with a quick pass through `argparse.ArgumentParser` and use that to initialize the `BlogClient`.

Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: Right! Updated `client.py` to take in CLI arguments.

File: signalfx-tracing/signalfx-python-tracing/django/README.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+```bash
+$ pip install -r requirements.txt
+# install supported tracer and library instrumentation
+$ sfx-py-trace-bootstrap
+$ cd blog
+$ python manage.py makemigrations posts
+$ python manage.py migrate
+$ python manage.py createsuperuser --username <USERNAME> --email <EMAIL>
+$ python manage.py runserver localhost:<PORT>
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/README.md]: `sfx-py-trace`?

File: signalfx-tracing/signalfx-python-tracing/django/README.md [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+```
+The client's service name may be set in `create_tracer()` as follows:
+```python
+
+create_tracer(service_name='<APP_NAME>')
+```
+If not set, the default service name is: `SignalFx-Tracing`<br/><br/>
+ 
+### Running the example
+If necessary, navigate to the folder containing `manage.py` (in this example:`django/blog`) and run:
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/README.md]: When would this not be necessary?

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
+# IMPORTANT:
+# The only requirement for auto-instrumentation is to add
+# 'signalfx_tracing' to installed apps.
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py]: nit: this is not entirely accurate since the tracer isn't created by this process (you still have to use sfx-py-trace or specify the tracer callable.  I'd say something like "To enable auto-instrumentation of Django apps, you must add the..."

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/urls.py [hunk 15]
```diff
@@ -0,0 +1,31 @@
+"""blog URL Configuration
+
+The `urlpatterns` list routes URLs to views. For more information please see:
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/blog/blog/urls.py]: I'd remove this tutorial boilerplate

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/wsgi.py [hunk 16]
```diff
@@ -0,0 +1,16 @@
+"""
+WSGI config for blog project.
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/blog/blog/wsgi.py]: same boilerplate removal

File: signalfx-tracing/signalfx-python-tracing/django/client.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+# Create and configure logger
+logging.basicConfig(level='DEBUG',
+                    format='%(asctime)s %(message)s',)
+
+
+# create an instance of the SignalFx tracer
+tracer = create_tracer(service_name='BlogClient')
+
+# Setting requests_config.propagate to 'True'  ensures
+# that the relational information between traces is maintained
```
Comment (rmfitzpatrick) [inline signalfx-tracing/signalfx-python-tracing/django/client.py]: "between spans" (Traces should be discrete and isolated events)

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py (hunk 14 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/blog/settings.py]: Correct, thanks. a bunch! Updated!

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/urls.py (hunk 15 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/blog/urls.py]: OK. Removed.

File: signalfx-tracing/signalfx-python-tracing/django/blog/blog/wsgi.py (hunk 16 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/blog/blog/wsgi.py]: Removed. Thanks!

File: signalfx-tracing/signalfx-python-tracing/django/client.py (hunk 17 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/client.py]: Thanks. Corrected!

File: signalfx-tracing/signalfx-python-tracing/django/README.md (hunk 12 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/README.md]: Yes! Corrected.

File: signalfx-tracing/signalfx-python-tracing/django/README.md (hunk 13 above)
Comment (sm7x) [reply signalfx-tracing/signalfx-python-tracing/django/README.md]: If the user is already in the folder 😀 . At this point in time, they probably are.... Re-worded it though, to remove any assumptions. 😁
